### ISSUE TITLE
Device mismatch in PreferenceModel when discounting

### ISSUE THREAD

**Issue by rk1a** [HumanCompatibleAI/imitation#619]
## Bug description
When training on preferences with a discounted sum of rewards the discounts are always on CPU, but the rewards from the reward model can be on GPU.

## Steps to reproduce
Run this on a machine with GPU:
`python src/imitation/scripts/train_preference_comparisons.py with preference_model_kwargs.discount_factor=0.99 fast`

## Environment
- Operating system and version: Ubuntu 20.04.5 LTS
- Python version: 3.8.13

**Comment by AdamGleave** [comment]
Not sure I quite understand the problem here. Is this a performance issue, or a correctness issue? Do you see an error message?The "discounts are always on CPU" -- the discount is a scalar float, so it seems appropriate it's on CPU, I assume you're referring to the discounted returns.

**Comment by rk1a** [comment]
It's a bug. Here is the error I get:
```
  File "src/imitation/scripts/train_preference_comparisons.py", line 257, in train_preference_comparisons
    results = main_trainer.train(
  File "/home/imitation/src/imitation/algorithms/preference_comparisons.py", line 1722, in train
    self.reward_trainer.train(self.dataset, epoch_multiplier=epoch_multiplier)
  File "/home/imitation/src/imitation/algorithms/preference_comparisons.py", line 1130, in train
    self._train(dataset, epoch_multiplier)
  File "/home/imitation/src/imitation/algorithms/preference_comparisons.py", line 1256, in _train
    loss = self._training_inner_loop(
  File "/home/imitation/src/imitation/algorithms/preference_comparisons.py", line 1316, in _training_inner_loop
    output = self.loss.forward(fragment_pairs, preferences, self._preference_model)
  File "/home/imitation/src/imitation/algorithms/preference_comparisons.py", line 1068, in forward
    probs, gt_probs = preference_model(fragment_pairs)
  File "/home/miniconda3/envs/imitation/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1190, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/imitation/src/imitation/algorithms/preference_comparisons.py", line 447, in forward
    probs[i] = self.probability(rews1, rews2)
  File "/home/imitation/src/imitation/algorithms/preference_comparisons.py", line 515, in probability
    returns_diff = (discounts * (rews2 - rews1)).sum(axis=0)
RuntimeError: Expected all tensors to be on the same device, but found at least two devices, cuda:0 and cpu!
```

I found that it can be fixed by moving the `discounts` tensor to the device of `rews1`.

**Comment by AdamGleave** [comment]
Thanks for the report, I can replicate the issue. @taufeeque9 can you look into a fix for this?

**Comment by AdamGleave** [comment]
Actually @mifeet can you look at this?

**Comment by mifeet** [comment]
I'm looking into it

**Comment by AdamGleave** [comment]
Closed in #636
